### PR TITLE
Add complete types to take all allowed keys into account

### PR DESCRIPTION
**Pull request by Viicos**
Fixes https://github.com/jpadilla/pyjwt/issues/602, https://github.com/jpadilla/pyjwt/issues/848, https://github.com/jpadilla/pyjwt/issues/856, https://github.com/jpadilla/pyjwt/issues/864#issuecomment-1458420475

I think this is the best we can get, without having to make use of `TypeVar`/`Generic` for algorithms. Had to tweak some parts to fix `mypy` errors.

One possible thing (as stated in https://github.com/jpadilla/pyjwt/pull/843) could be to add overloads to `encode`. I'll let you decide on this one, as it can hurt code readability.
Regarding the `decode` method, it can't be done as a list of allowed algorithms is passed.

### CODE REVIEW

**Review by auvipy — CHANGES_REQUESTED** [review]
can you please check why CI is failing?

**Comment by Viicos** [comment]
> can you please check why CI is failing?

```
  self = <jwt.api_jwt.PyJWT object at 0x7f36e4b70610>
  payload = {'claim': 'insanity', 'exp': 1681192610, 'iss': 'jeff'}
  now = 1681192613.000076, leeway = 3
  
      def _validate_exp(
          self,
          payload: dict[str, Any],
          now: float,
          leeway: float,
      ) -> None:
          try:
              exp = int(payload["exp"])
          except ValueError:
              raise DecodeError("Expiration Time claim (exp) must be an" " integer.")
      
          if exp <= (now - leeway):
  >           raise ExpiredSignatureError("Signature has expired")
  E           jwt.exceptions.ExpiredSignatureError: Signature has expired
  
  jwt/api_jwt.py:304: ExpiredSignatureError
```

Seems like this could fail randomly

File: tests/test_algorithms.py [hunk 1]
```diff
@@ -767,12 +767,6 @@ def test_okp_ed25519_should_reject_non_string_key(self):
         with open(key_path("testkey_ed25519.pub")) as keyfile:
             algo.prepare_key(keyfile.read())
 
-    def test_okp_ed25519_should_accept_unicode_key(self):
```
Comment (auvipy) [inline tests/test_algorithms.py:770]: can you please describe why this test is removed?

**Review by auvipy — CHANGES_REQUESTED** [review]

File: tests/test_algorithms.py (hunk 1 above)
Comment (Viicos) [reply tests/test_algorithms.py:770]: It's a duplicate of

https://github.com/jpadilla/pyjwt/blob/ba726444a6cee75af59feb8ea08294d0ac89bedb/tests/test_algorithms.py#L758-L765

L764-765

In fact I think it's better to move the two valid checks from `test_okp_ed25519_should_reject_non_string_key` to this removed `test_okp_ed25519_should_accept_unicode_key`. Tell me if you want me to do it this way

**Comment by Viicos** [comment]
Most of the remaining `mypy` errors are due to the fact that `alg.prepare_key`/`alg.from_jwk` can return either a public or a private key. Do you want to keep mypy running for tests and have an explicit assignment for the return types of these methods, or just drop `mypy` for tests?

**Comment by auvipy** [comment]
> Do you want to keep mypy running for tests and have an explicit assignment for the return types of these methods

yes

**Review by auvipy — APPROVED** [review]

**Pull request merged**